ctapipe/core/traits.py

```diff
@@ -138,22 +139,51 @@ def has_traits(cls, ignore=("config", "parent")):
     return bool(set(cls.class_trait_names()) - set(ignore))
 
 
+class TelescopePatternList(UserList):
+    """
+    Representation for a list of telescope pattern tuples. This is a helper class
+    used  by the Trait TelescopeParameter as its value type
+    """
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._lookup = None
+        self._old_getitem = self.__getitem__
+        self._subarray = None
+
+    @property
+    def tel(self):
+        """ access the value per telescope_id, e.g. `param.tel[2]`"""
+        if self._lookup:
+            return self._lookup
+        else:
+            raise RuntimeError("No TelescopeParameterLookup was registered")
+
+    def attach_subarray(self, subarray: "ctapipe.instrument.SubarrayDescription"):
+        """
+        Register a SubarrayDescription so that the user-specified values can be
+        looked up by tel_id. This must be done before using the `.tel[x]` property
+        """
+        self._subarray = subarray
+        self._lookup.attach_subarray(subarray)
+
+
 class TelescopeParameterLookup:
-    def __init__(self, telecope_parameter_list):
+    def __init__(self, telescope_parameter_list):
         """
         Handles the lookup of corresponding configuration value from a list of
         tuples for a telid.
 
         Parameters
         ----------
-        telecope_parameter_list : list
+        telescope_parameter_list : list
             List of tuples in the form `[(command, argument, value), ...]`
         """
-        self._telecope_parameter_list = telecope_parameter_list
+        self._telescope_parameter_list = telescope_parameter_list
```

Comment:
`TelescopePatternList` holds a reference to its `TelescopeParameterLookup` and the other way around, so these objects will never be garbage collected.

Give the Lookup a `weakref` or avoid the mutual reference?

Reply:
This is perhaps leftover from the original implementation - not sure we need to keep the telescope_parameter_list... let me check.  Otherwise I'll use a WeakRef

Reply:
seems it's only really used to get the global values in the case that `attach_subarray()` has not been run, but it would require quite some refactoring to make that work without it, so I think I'll just make it a weakref and be done (hopefully this will all be replaced with a better config system soon anyhow)

Reply:
I decided just to do a deepcopy() of the TelescopeList inside TelescopeLookup. That should work, I think.

Reply:
seems that what was making the "retain subarray" between changes functionality work, which is somewhat complex. I'll see if I can fix it, but I think it will require some large refactoring from what was implemented before. Perhaps that's not a necessary feature?  @watsonjj - you added that feature - is that really important?

Reply:
I don't remember if anything depends on this behaviour. It just seemed unintuitive that the subarray is forgotten when you change the values of the traitlet.

Reply:
> so these objects will never be garbage collected.

Is that really true? From what I can find, they won't be freed immediately, as they will never reach a reference count of zero. But the garbage collector itself will periodically free circularly-referenced objects.

Reply:
Ah, yes. In fact, the whole point of the gc is circular references, as everything else is handled by ref counting.

Reply:
I'm not sure that the functionality broke because of the circular ref.  I think it's just because now if any change is made, a new TelescopeList is generated by validate(), and thus the link to the TelescopeParametersLookup is lost.  Strangely, it did work after I first refactored, so I must have changed something recently. Anyhow, I'll see if I can get it working again.